### PR TITLE
Use ruff (GSI-339)

### PR DESCRIPTION
**Pull request by TheByronHimes**
:dog: Using ruff to replace some of our code quality tools.
---
>The initial branch for these changes ended up with a lot of merge conflicts since the pip-tools changes occurred in the interim. Rather than try to untangle everything I decided to start over. 
---
Ruff is a python linter written in rust that runs much, [much ](https://beta.ruff.rs/docs/) faster than pylint. 
Ruff is not a drop-in replacement for pylint, but it implements several dozen of pylint's rules (see [here](https://beta.ruff.rs/docs/rules/#pylint-pl) for the list and [here](https://github.com/astral-sh/ruff/issues/970) for a comparison)). Looking at the parity tracking, ruff seems to implement the most important rules from pylint. 
It can, however, be used as a [drop-in replacement](https://beta.ruff.rs/docs/faq/#how-does-ruff-compare-to-flake8) for flake8. 

### Changes
- Pylint, flake8, bandit, and isort have been replaced in the common dev dependencies by ruff.
- Pylint, flake8, bandit, and isort pre-commit hooks have been replaced by ruff's pre-commit hook.
- Pylint, flake8, bandit, and isort github actions have been replaced by the ruff github action. 
- The .ruff.toml config file has been added and the pylint/flake8 config files have been removed.
- Fixed: added `update_lock.py` to `.static_files` (was missed in [previous PR](https://github.com/ghga-de/microservice-repository-template/pull/160))
- Added `scripts/update_hook_revs.py` to keep the ruff/black/mypy pre-commit hook versions in line with the dependencies in the lock file. This is also added as a first-step pre-commit hook run with --check.
  - Since the dev dependency versions are now uncapped, it would be difficult to enforce versions centrally.
- Expanded the settings in `.ruff.toml` (thanks @Cito) to include flake8-bugbear, flake8-bandit, ruff's rules from pylint, isort, pyupgrade, and more.

### CODE REVIEW

File: .github/workflows/static_code_analysis.yaml [hunk 1]
```diff
@@ -15,24 +15,14 @@ jobs:
       - uses: pre-commit/action@v3.0.0
         env:
           SKIP: no-commit-to-branch
+      - name: ruff
+        uses: chartboost/ruff-action@v1
       - name: black
         run: |
           black --check .
-      - name: isort
```
Comment (KerstenBreuer) [inline .github/workflows/static_code_analysis.yaml:21]: I saw that ruff also checks the import sorting. However, will it also correct it? I.e. is it acting as a linter or a formatter here? If the it's only linting, we might still consider using isort.

File: .github/workflows/static_code_analysis.yaml [hunk 2]
```diff
@@ -15,24 +15,14 @@ jobs:
...
       - name: mypy
         run: |
           mypy .
-      - name: pylint
-        run: |
-          pylint "${{ steps.common.outputs.MAIN_SRC_DIR }}"
-      - name: flake8
-        run: |
-          flake8 --config .flake8
-      - name: bandit
```
Comment (KerstenBreuer) [inline .github/workflows/static_code_analysis.yaml:33]: Is ruff also implementing the checks of bandit?

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -7,6 +7,16 @@ default_language_version:
 minimum_pre_commit_version: 3.0.0
 
 repos:
+  - repo: local
```
Comment (KerstenBreuer) [inline .pre-commit-config.yaml:10]: Very nice indeed

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -36,33 +46,17 @@ repos:
       - id: debug-statements
       - id: debug-statements
       - id: debug-statements
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.0.290
+    hooks:
+      - id: ruff
+        args: [--fix]
```
Comment (KerstenBreuer) [inline .pre-commit-config.yaml:53]: Don't we need this (according to the documentation: https://docs.astral.sh/ruff/usage/#pre-commit):

```suggestion
      args: [ --fix, --exit-non-zero-on-fix ]
```

File: .ruff.toml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    "RUF", # ruff
+    "SIM", # flake8-simplify
+]
+
+fixable = [
+    "UP", # e.g. List -> list
+    "I", # sort imports
+]
+
+target-version = "py39"
```
Comment (KerstenBreuer) [inline .ruff.toml:38]: Cool, I just saw that ruff also replaces pyupgrade :)

File: scripts/update_hook_revs.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    """Compare configured pre-commit hooks with the installed dependencies. For the set
+    that overlap (e.g. `black`, `mypy`, `ruff`, etc.), make sure the versions match.
+    If running with `--check`, exit with status code 1 if anything is outdated.
+    If running without `--check`, update `.pre-commit-config.yaml` as needed.
+    Any invalid flags are ignored.
+    """
+    args = sys.argv[1:]
+    running_pre_commit = "--check" in args
+
+    dependency = re.compile(r"([^=\s]+)==([^\s]*?)\s")
```
Comment (KerstenBreuer) [inline scripts/update_hook_revs.py:38]: Maybe better name dependency_pattern

File: scripts/update_hook_revs.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+def main():
+    """Compare configured pre-commit hooks with the installed dependencies. For the set
+    that overlap (e.g. `black`, `mypy`, `ruff`, etc.), make sure the versions match.
+    If running with `--check`, exit with status code 1 if anything is outdated.
+    If running without `--check`, update `.pre-commit-config.yaml` as needed.
+    Any invalid flags are ignored.
+    """
+    args = sys.argv[1:]
+    running_pre_commit = "--check" in args
```
Comment (KerstenBreuer) [inline scripts/update_hook_revs.py:36]: Please have a look at how we typically write the update scripts using our wrapper around typer:
https://github.com/ghga-de/microservice-repository-template/blob/main/scripts/update_lock.py#L185
https://github.com/ghga-de/microservice-repository-template/blob/main/scripts/update_lock.py#L257

We also have methods for reporting success or failures to the user:
https://github.com/ghga-de/microservice-repository-template/blob/main/scripts/update_lock.py#L248

Would be great if all of the scripts follow the same pattern.
https://github.com/ghga-de/microservice-repository-template/blob/main/scripts/update_lock.py#L248

**Review by KerstenBreuer — CHANGES_REQUESTED** [review]
Looks great. As always, an outstanding PR description, that really helps and serves as a good documentation if later questions come up.

Thanks also for te prep-commit update scripts, that really handy in deed.

Just some cosmetics.

File: .github/workflows/static_code_analysis.yaml (hunk 2 above)
Comment (TheByronHimes) [reply .github/workflows/static_code_analysis.yaml:33]: Yes, through implementing the rules of flake8-bandit.

Comment (KerstenBreuer) [reply .github/workflows/static_code_analysis.yaml:33]: Nice, getting more and more convinced of ruff, hope it also works as expected

File: .pre-commit-config.yaml (hunk 4 above)
Comment (TheByronHimes) [reply .pre-commit-config.yaml:53]: Ah yeah I see that now.

File: scripts/update_hook_revs.py (hunk 6 above)
Comment (TheByronHimes) [reply scripts/update_hook_revs.py:38]: Agreed and changed

File: scripts/update_hook_revs.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+import re
+import sys
+from pathlib import Path
+
+REPO_ROOT_DIR = Path(__file__).parent.parent.resolve()
+PRE_COMMIT_CFG_PATH = REPO_ROOT_DIR / ".pre-commit-config.yaml"
+LOCK_FILE_PATH = REPO_ROOT_DIR / "requirements-dev.txt"
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply scripts/update_hook_revs.py]: see below

File: scripts/update_hook_revs.py (hunk 7 above)
Comment (TheByronHimes) [reply scripts/update_hook_revs.py]: Main function split out and typer/cli.py used as of [13efb09](https://github.com/ghga-de/microservice-repository-template/pull/161/commits/13efb09e24bd7604d5d7b051809bd39c2061d8f2)

File: scripts/update_hook_revs.py (hunk 6 above)
Comment (TheByronHimes) [reply scripts/update_hook_revs.py]: Fixed

File: .pre-commit-config.yaml (hunk 4 above)
Comment (TheByronHimes) [reply .pre-commit-config.yaml]: Fixed

File: .ruff.toml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    "C90", # McCabe Complexity
+    "F", # pyflakes codes
+    "I", # isort
+    "S", # flake8-bandit
+    "B", # flake8-bugbear
+    "N", # pep8-naming
+    "UP", # pyupgrade
+    "PL", # pylint
+    "RUF", # ruff
+    "SIM", # flake8-simplify
```
Comment (Cito) [inline .ruff.toml:30]: What about D = pydocstyle?

File: scripts/update_hook_revs.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        if match:
+            package, version = match.groups()
+            dependencies[package] = version
+
+    return dependencies
+
+
+def get_repl_value(match, dependencies: dict[str, str], outdated_hooks: list[str]):
+    """Look up pre-commit hook id in list of dependencies. If there's a match, update
+    return the version stored in the dictionary"""
```
Comment (Cito) [inline scripts/update_hook_revs.py:51]: Something is wrong with the grammar of the docstring.

**Review by Cito — DISMISSED** [review]
Nice work. Sorry, could only do a superficial review.

File: .ruff.toml [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    ".git",
+    ".devcontainer",
+    "__pycache__",
+    "build",
+    "dist",
+]
+
+ignore = [
+    "E",
+    "W",
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .ruff.toml:11]: @Cito I was following our previous flake8 configuration:
```
ignore = E, W
    # ignore all style checks from pycodestyle
    # as they are already checked by black
```

[the remark this answers is not in the document]
Comment (Cito) [reply .ruff.toml:11]: Ok, makes sense.

[the remark this answers is not in the document]
Comment (Cito) [reply .ruff.toml:11]: Maybe add that comment from the flake8 config to the toml file then.

File: scripts/update_hook_revs.py (hunk 10 above)
Comment (TheByronHimes) [reply scripts/update_hook_revs.py]: Does it update? Does it return? Who knows? 
Edited it in the commit below.

**Review by Cito — DISMISSED** [review]
Can you still add comments to all entries in the "ignore" list?

**Review by Cito — APPROVED** [review]

**Review by KerstenBreuer — APPROVED** [review]

**Pull request merged**